Chrispresso/08_05_2019.md

```diff
@@ -0,0 +1,125 @@
+* Finished verifying that my genetic algorithm was working. Commit can be seen here: https://github.com/Chrispresso/PyGenoCar/commit/774ae973266b8eae30044737450b1732d4615ce9
+
+
+Because of how I abstracted away some of the GA, I'm able to create different types of GAs based on the problem. For example:
+
+~~~~
+import numpy as np
+from typing import Optional
+from genetic_algorithm.individual import Individual
+from genetic_algorithm.population import Population
+from genetic_algorithm.selection import *
+from genetic_algorithm.crossover import simulated_binary_crossover as SBX
+from genetic_algorithm.mutation import gaussian_mutation
+from matplotlib import pyplot as plt
+
+
+### Different functions to test based off 
+### https://en.wikipedia.org/w/index.php?title=Test_functions_for_optimization&oldid=787014841
+
+# func_to_maximize = lambda v,w,x,y,z: -abs(abs(v**3) + (w)**2 + (x-5)**2 + (y-7)**2 + (z + 9)**2)
+# func_to_maximize = lambda x,y: -abs(2*x**2 - 1.05*x**4 + ((x**6)/6) + x*y + y**2)  # 3 hump camel func
+# func_to_maximize = lambda x,y: -abs( .26*(x**2 + y**2) - .48*x*y )
+# func_to_maximize = lambda x,y: -abs( -20*np.exp(-.2 * np.sqrt(.5*(x**2 + y**2))) -np.exp(.5 * (np.cos(2*np.pi*x) + np.cos(2*np.pi*y))) + np.exp(1) + 20 )  # Ackley func - doesnt find (0,0)
+# func_to_maximize = lambda x,y: -abs( -np.cos(x)*np.cos(y)*np.exp( -((x-np.pi)**2 + (y-np.pi)**2) ) )  # Easom func
+func_to_maximize = lambda x,y: -abs( -(y + 47)*np.sin( np.sqrt( abs( (x/2) + (y+47) ) ) ) - x*np.sin( abs(x - (y+47)) ) )
+
+
+class ValIndividual(Individual):
+    def __init__(self, chromosome: Optional[np.ndarray] = None):
+        super().__init__()
+        if not chromosome is None:

```

Comment:
Can this be 'if chromosome' or would that not work?

Reply:
In this case no. Unfortunately `np.ndarray` considers the truth value of an array with more than one element to be ambiguous. So if the array is given it will throw an error. I could do `if chromosome is not None` which might make it easier to read.